### ISSUE TITLE
finalization semantic in mercury

### ISSUE THREAD

**Issue by JohnPJenkins** [mercury-hpc/mercury#53]
While doing some mercury resource cleanup I've been thinking about how to resolve races and gracefully finalize. Consider multiple services built atop mercury, sharing the mercury endpoint. Now, service X wants to finalize and clean itself up. One could imagine this order of operation:
1) turn away any new RPCs (exposed by the service to clients)
2) drain the pending RPC pool for this service, executing (or cancelling) any remaining RPCs
3) performing service-specific cleanup work (which may involve calling service internal RPCs)
3) de-registering the RPCs

This gives us a seemingly reasonable 2-step shutdown semantic, in which we first disable all RPCs associated with the service, then remove them entirely from the system. There are still potential issues (RPCs that call other RPCs, for instance), but the listed semantic seems like a good start.

Assuming for the moment that these semantics make sense :), there needs to be a Mercury API to facilitate, as there's no way to inspect on a per-RPC basis the state of pending operations, queue sizes, etc. One could imagine the following set of calls:

```C
// Disable the receipt of any newly-received RPCs with the given ID.
// If cancel_pending is true, then any pending RPCs will be cancelled and notification sent to clients. Otherwise, pending RPCs will be triggered as normal.
// RPCs of the given type executing concurrently with this call will be unaffected by this call.
// This call should only be done for listening Mercury endpoints - it is nonsensical for RPC clients.
hg_return_t HG_Disable(hg_class_t *hg_class, hg_id_t id, bool cancel_pending);
// TODO: multi-id version of this?

// Register an RPC-like callback that is executed through HG_Trigger when the pending set of RPCs for the given disabled RPC has been cleared from the system.
// The input struct for the resulting RPC handle is of type hg_id_t, representing the input RPC id.
// As this "RPC" function is only called internally, there are no pending RPCs to wait on, and so it can be directly de-registered once no longer needed.
// It is recommended to register the function prior to calling HG_Disable, as registering afterwards can cause race conditions when Mercury progression is concurrent with the caller.
// By default, no notification will be given after the pending RPCs for a disabled RPC are cleared.
hg_id_t HG_Register_disable_cleared(hg_class_t *hg_class, hg_id_t in_id, hg_rpc_cb_t cb);
// TODO: multi-id version of this?

// TODO: combine the disable and registration call?

// De-register an RPC, removing it and all associated metadata. There must not be any pending RPCs of the given ID when this call is made.
hg_return_t HG_Deregister(hg_class_t *hg_class, hg_id_t id);
```

Is this a reasonable assessment / solution?

**Comment by soumagne** [comment]
I see, yes I like the idea of giving more control to the upper layer to finalize things and yes what you describe is currently done internally only within the HG_Finalize call. I'm not really sure though what you would gain by deregistering the RPC? Do you intend to re-use the same HG class later on? What I would do in this case is same as you describe 1, 2 and then call Finalize().

**Comment by JohnPJenkins** [comment]
The use case for de-registration I'm thinking of is a long-running set of mercury processes by which services come and go, i.e. the class ends up getting re-used. De-registering allows bounding resource use over time and reuse of the RPC ID, which is useful if you buy into the idea of mercury being used in a highly dynamic environment (certainly not something we're worrying about in the short term, but plausible in the longer term I think?).

There's also potentially a distinction to be made between a non-existent RPC (one that's been de-registered) and a "shutting down" RPC (in which some set of processes have HG_Disable'd RPCs). Not sure how important this distinction is though...

**Comment by soumagne** [comment]
OK. And it's also true that we have a HG_Register() call so it probably just makes sense to provide a Deregister() call :)
